ecosystem/indexer-grpc/transaction-filter/src/traits.rs

```diff
@@ -0,0 +1,202 @@
+// Copyright (c) Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::errors::FilterError;
+use serde::Serialize;
+use std::fmt::Debug;
+
+/// Simple trait to allow for filtering of items of type T
+pub trait Filterable<T>
+where
+    Self: Debug + Serialize,
+{
+    /// Whether this filter is correctly configured/initialized
+    /// Any call to `validate_state` is responsible for recursively checking the validity of any nested filters *by calling `is_valid`*
+    /// The actual public API is via `is_valid` which will call `validate_state` and return an error if it fails, but annotated with the filter type/path
+    fn validate_state(&self) -> Result<(), FilterError>;
+
+    /**
+     * This is a convenience method to allow for the error to be annotated with the filter type/path at each level
+     * This is the public API for checking the validity of a filter!
+     * Example output looks like:
+     * ```text
+     *   FilterError: This is a test error!.
+     *   Trace Path:
+     *   transaction_filter::traits::test::InnerStruct:   {"a":"test"}
+     *   core::option::Option<transaction_filter::traits::test::InnerStruct>:   {"a":"test"}
+     *   transaction_filter::traits::test::OuterStruct:   {"inner":{"a":"test"}}
+     *  ```
+     **/
```

Comment:
Needs to be `///` for it to be a doc comment. Maybe intentional that it isn't one? Seems like it should be though.